FreeRTOS/FreeRTOS-Plus-TCP
# Preparing library to use loopback device

This PR prepares the IP-stack to use loopback addresses `127.0.0.0` and `::1`.

The issue was raised on the FreeRTOS forum [here](https://forums.freertos.org/t/freertos-tcp4-0-0-init-and-loopback-problems/18034) by [ozanagma](https://forums.freertos.org/u/ozanagma).

Description
-----------

● Routing.c :

It now recognises a new IP type for loopback: "eIPv6_Loopback"

Removed a minor bug in `FreeRTOS_AddNetworkInterface()`, and also `FreeRTOS_AddEndPoint()`.

When adding the following interfaces:
~~~
    FreeRTOS_AddNetworkInterface( &iface_a );
    FreeRTOS_AddNetworkInterface( &iface_b );
    FreeRTOS_AddNetworkInterface( &iface_a );
    FreeRTOS_AddNetworkInterface( &iface_c );
~~~
In the above example, "iface_b" will become an orphan because of this piece of code:
~~~
    if( pxInterface != NULL )
    {
        pxInterface->pxNext = NULL;
        pxInterface->pxEndPoint = NULL;
~~~

● FreeRTOS_IPv6.h :

Added `xBadIPv6Loopback()`
Added `xIsIPv6Loopback()`

● FreeRTOS_IPv4.h :

Added `xIsIPv4Loopback()`
Added `xBadIPv4Loopback()`

Communication with a loopback device always takes place within a system, and from a loopback IP address to another loopback IP.

Here is how I implemented the XOR operation:
~~~
    BaseType_t xReturn = pdFALSE;
    if( pxEndPoint != NULL )
    {
        BaseType_t x1 = ( xIsIPv4Loopback( pxIPHeader->ulDestinationIPAddress ) != 0 ) ? pdTRUE : pdFALSE;
        BaseType_t x2 = ( xIsIPv4Loopback( pxIPHeader->ulSourceIPAddress ) != 0 ) ? pdTRUE : pdFALSE;

        if( x1 != x2 )
        {
            /* Either the source or the destination address is a loopback address. */
            xReturn = pdTRUE;
        }
    }
~~~

● FreeRTOS_ND.c :

The function `FreeRTOS_CreateIPv6Address()` will allow prefix address length of 128 bits, which is necessary to allow IPv6 loopback addresses.

● FreeRTOS_IP.c :

and the most important change:

~~~diff
-    else if( ( pxUDPPacket->xEthernetHeader.usFrameType == ipIPv4_FRAME_TYPE ) &&
-             ( ipFIRST_LOOPBACK_IPv4 <= ( FreeRTOS_ntohl( pxUDPPacket->xIPHeader.ulDestinationIPAddress ) ) ) &&
-             ( ( FreeRTOS_ntohl( pxUDPPacket->xIPHeader.ulDestinationIPAddress ) ) < ipLAST_LOOPBACK_IPv4 ) )
+    else if( ( pxUDPPacket->xEthernetHeader.usFrameType == ipIPv4_FRAME_TYPE ) &&
+             ( xBadIPv4Loopback( &( pxUDPPacket->xIPHeader ) ) != pdFALSE ) )
+    {
+        /* The local loopback addresses must never appear outside a host. See RFC 1122
+         * section 3.2.1.3. */
+        eReturn = eReleaseBuffer;
+    }
~~~

● FreeRTOS_IPv6.c :

a similar changes as in FreeRTOS_IP.c here above.


Test Steps
-----------
In my next PR I will present the loopback network interface.

Also I developed a module that starts 4 tasks to do TCPv4, TCPv6, UDPv4, and UDPv6. Each task will create to sockets that will exchange data through the loopback device.


Checklist:
----------



Related Issue
-----------



By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
